In [1]:
import pandas as pd

df_facebook = pd.read_csv('./case estágio de dados/last-30d-facebook.csv')

df_facebook

Campanha    Base Usuários Evento Otimizador         Mídia      Gasto  \
0         A      Recomprador          Purchase  facebook ads   R$ 5.000   
1         B      Recomprador          Purchase  facebook ads  R$ 15.000   
2         C  Primeira Compra          Purchase  facebook ads   R$ 9.000   
3         D  Primeira Compra          Purchase  facebook ads  R$ 12.900   
4         E  Primeira Compra          Purchase  facebook ads   R$ 1.000   
5         F  Primeira Compra          Purchase  facebook ads     R$ 400   
6         G  Primeira Compra          Purchase  facebook ads     R$ 250   
7         H  Primeira Compra          Purchase  facebook ads  R$ 10.000   
8         I  Primeira Compra          Purchase  facebook ads  R$ 12.000   
9         J  Primeira Compra          Purchase  facebook ads  R$ 21.500   
10        K      Recomprador          Purchase  facebook ads   R$ 1.050   
11        L  Primeira Compra          Purchase  facebook ads     R$ 100   

   Investimento alocado   Clicks Impressões  Sessões  Conversão    Receita  \
0              R$ 4.500   20.000  2.000.000    19.00        133  R$ 13.300   
1             R$ 20.000   30.000  3.333.333    28.50        143  R$ 21.375   
2              R$ 8.000   60.000  7.500.000    57.00        342  R$ 18.810   
3             R$ 13.000   25.800  1.720.000    24.51        123  R$ 12.010   
4              R$ 1.500    5.000    505.051     4.75         71   R$ 9.476   
5              R$ 2.500  444.000     17.778   422.00         11   R$ 2.639   
6                R$ 600  500.000    333.333   475.00          7   R$ 2.494   
7             R$ 12.000   30.000  5.000.000    25.00        120   R$ 9.120   
8             R$ 12.000   15.000  1.666.667    12.00        187   R$ 8.424   
9             R$ 20.000   37.504  3.261.217    32.00        256  R$ 15.104   
10             R$ 2.500    1.500     60.000     1.40         56  R$ 14.000   
11             R$ 1.000   80.000      4.211    81.00          2     R$ 911   

    Primeira compra  
0                20  
1                20  
2               250  
3                40  
4                25  
5                 2  
6                 5  
7                75  
8                80  
9               225  
10                2  
11                2

In [5]:
# função para converter para int o valor em R$

def delete_real(data: pd.Series) -> pd.DataFrame:
    delete = "R$ ."

    for i in range(len(data)):
        for x in delete:
            if x in data[i]:
                data[i] = data[i].replace(x, "")
    
    data = pd.DataFrame((data.astype(int)))

    return data

df_gasto = delete_real(df_facebook['Gasto']).to_dict()
df_invest = delete_real(df_facebook['Investimento alocado']).to_dict()
df_receita = delete_real(df_facebook['Receita']).to_dict()

# cria um novo DataFrame com os dados convertidos para facilitar a análise e os cálculos

df_total = pd.DataFrame(data=(df_gasto["Gasto"], df_invest['Investimento alocado'], df_receita['Receita'])).transpose().rename(columns={0: 'Gasto', 1: 'Investimento alocado', 2: 'Receita'})

In [3]:
list_result = []    # lista para armazenar os resultados de cada campanha em ordem

# laço para calcular a relação de gasto e receita em cada uma das campanhas

for i in range(len(df_total)):
    porcentagem = (df_total['Receita'][i] * 100) / df_total['Gasto'][i]

    list_result.append(f"{porcentagem:.2f} %")

list_result

['266.00 %',
 '142.50 %',
 '209.00 %',
 '93.10 %',
 '947.60 %',
 '659.75 %',
 '997.60 %',
 '91.20 %',
 '70.20 %',
 '70.25 %',
 '1333.33 %',
 '911.00 %']

In [4]:
# adiciona uma nova linha da relação de gasto e receita no dataframe original

df_facebook['Relação Gasto e Receita'] = list_result

df_facebook

Campanha    Base Usuários Evento Otimizador         Mídia  Gasto  \
0         A      Recomprador          Purchase  facebook ads   5000   
1         B      Recomprador          Purchase  facebook ads  15000   
2         C  Primeira Compra          Purchase  facebook ads   9000   
3         D  Primeira Compra          Purchase  facebook ads  12900   
4         E  Primeira Compra          Purchase  facebook ads   1000   
5         F  Primeira Compra          Purchase  facebook ads    400   
6         G  Primeira Compra          Purchase  facebook ads    250   
7         H  Primeira Compra          Purchase  facebook ads  10000   
8         I  Primeira Compra          Purchase  facebook ads  12000   
9         J  Primeira Compra          Purchase  facebook ads  21500   
10        K      Recomprador          Purchase  facebook ads   1050   
11        L  Primeira Compra          Purchase  facebook ads    100   

   Investimento alocado   Clicks Impressões  Sessões  Conversão Receita  \
0                  4500   20.000  2.000.000    19.00        133   13300   
1                 20000   30.000  3.333.333    28.50        143   21375   
2                  8000   60.000  7.500.000    57.00        342   18810   
3                 13000   25.800  1.720.000    24.51        123   12010   
4                  1500    5.000    505.051     4.75         71    9476   
5                  2500  444.000     17.778   422.00         11    2639   
6                   600  500.000    333.333   475.00          7    2494   
7                 12000   30.000  5.000.000    25.00        120    9120   
8                 12000   15.000  1.666.667    12.00        187    8424   
9                 20000   37.504  3.261.217    32.00        256   15104   
10                 2500    1.500     60.000     1.40         56   14000   
11                 1000   80.000      4.211    81.00          2     911   

    Primeira compra Relação Gasto e Receita  
0                20                266.00 %  
1                20                142.50 %  
2               250                209.00 %  
3                40                 93.10 %  
4                25                947.60 %  
5                 2                659.75 %  
6                 5                997.60 %  
7                75                 91.20 %  
8                80                 70.20 %  
9               225                 70.25 %  
10                2               1333.33 %  
11                2                911.00 %

## A) Considerando a base de Facebook Ads, responda:

### 1. Como você realocaria o investimento entre as campanhas de Facebook Ads afim de otimizar o seu retorno sobre investimento?

De início, importei o dataset do facebook para a classe DataFrame do Pandas, tornando os dados mais facilmente interpretáveis.

Afim de interpretar melhor os dados para responder a questão, inseri uma nova coluna no dataframe para relacionar a receita gerada com o anúncio ao gasto da empresa com o mesmo, desta forma sendo mais fácil visualizar quais campanhas geraram receitas mais satisfatórias.

Desta forma, para gerar um melhor investimento nas campanhas, visualmente, seria realocar os investimentos de campanhas que a relação entre receita e gasto ficaram abaixo dos 100%, ou seja, que a receita não superou o gasto com o anúncio, para campanhas que a receita superou o gasto com a mesma, desta forma otimizando os gastos para gerar mais receita para a empresa.

**Porém, não seria uma realocação adequada somente analisar os dados da relação entre receita e gasto**, mas também analisar outras métricas, como as impressões, clicks, conversão e base de usuários, assim deve-se obter uma análise mais completa do caso. Seguindo essa lógica, iremos analisar essas métricas relacionanda-as para um melhor resultado.

De forma correta, teríamos que analisar os dados de maneira concisa, relacionando as colunas que possuem uma relação (indireta ou direta) para obter um melhor resultado na realocação.

Para realocar corretamente os investimentos, seria necessário distinguir quais campanhas estão atingindo seu objetivo, relacionado à coluna **"Base Usuários"**, ou seja, se uma campanha está atingindo o resultado, o público alvo desta campanha está comprando, é um ponto a ser considerado para investir nesta campanha. Analisando outro ponto, as **impressões e clicks** é uma métrica interessante a se analisar, pois além de trazer novos ou já compradores da empresa, elas ajudam a divulgar a marca, contribuindo para a socialização da empresa e conhecimento de várias pessoas sobre os produtos da marca, então são métricas interessantes a analisar para chegar a uma alocação de investimentos eficaz. Uma outra métrica que considero para realizar uma análise boa, é a conversão, que dita sobre quantas compras foram feitas por meio da campanha, quanto mais compras são realizadas por aquela campanha, mais interessante é o anúncio, ponto a ser considerado para a alocação de recursos.

Exemplificando o que foi dito a cima, irei realizar a análise de algumas campanhas para mostrar minha ideia de realocação dos recursos.

Analisando a campanha A, temos algumas métricas importantes que são colocadas abaixo:

Campanha | Base usuários | Impressões  | Clicks | Primeira compra | Conversão | Relação Gasto e Receita
-------  | ------------- | ----------- | ------ | --------------- | --------- | ---------
A        | Recomprador   | 2.000.000   | 20.000 | 20              | 133       | 266 %

É importante analisar primeiro que ela atingiu seu objetivo como público de comprador, mostrando que das compras realizadas, 113 foram de pessoas que voltaram a comprar da empresa. A campanha também obteve uma relação de gasto e receita boa, gerando lucro para a empresa, além de uma boa divulgação com as impressões e clicks. Desta forma, esta seria uma campanha que seu investimento poderia ser mantido para aumentar o público que retorna para comprar na empresa.

Agora analisando a campanha J:

Campanha | Base usuários   | Impressões  | Clicks | Primeira compra | Conversão | Relação Gasto e Receita
-------  | -------------   | ----------- | ------ | --------------- | --------- | ---------
J        | Primeira compra | 3.261.217   | 37.504 | 225             | 256       | 70.25 %

Esta campanha teve um investimento alto, porém comparada a seu gasto, sua receita foi baixa, desta forma teria que ser reformulada para gerar um melhor resultado. Mas por que não pausar esta campanha? Pois apesar de seu gasto ser alto e não trazer receitas tão boas quanto outras, ela gerou uma boa visualização da empresa e também teve uma boa quantidade de primeiras compras, ou seja, a função dela de atrair clientes foi cumprida, além também de ter um número alto de impressões e clicks no anúncio, desta forma aumentando o engajamento e repercursão da empresa. Então seria uma boa opção, realocar o investimento nela, reduzindo-o e alocando-o para outras campanhas, como a C, E, G e K, que foram boas campanhas e geraram bons resultados, tanto em divulgação da empresa como em receita.

Seria dessa forma, enfim, que eu realizaria a realocação dos investimentos para gerar um melhor resultado para a empresa.

### 2. Qual é a(s) principal(is) métrica(s) que você acredita que exemplifica o potencial de conversão de um anúncio? Baseado na sua resposta, quais campanhas pausaria?

Como supracitado, não é somente uma métrica que exemplifica o potencial de conversão, podem ser várias métricas que irão dizer a capacidade daquela campanha gerar conversões. Acredito que as principais métricas, presentes no dataset, que ditam sobre a capacidade de conversão são fundamentalmente as impressões, podendo ser ampliada sua análise por meio dos clicks, que também influenciam, pois quanto mais clicks em uma campanha, mais chances existem de um produto ser comprado. 

Porém, existem outras métricas que podem ser analisadas e que podem influenciar e muito na conversão, são as métricas do próprio anúncio, refiro-me as características do mesmo, como as cores utilizadas, a disposição das palavras na campanha, essas são métricas importantes para analisar corretamente a possibilidade de conversão do anúncio. Um anúncio que tenha características agradáveis e cores que combinam é mais influente para compra do que uma propaganda que não tenha uma disposição legal das suas caracteríticas. Evidencia-se, portanto, que uma análise somente dentro do dataset não seria suficiente para analisar o potencial de conversão, seria necessário passar por vários fatores.

As métricas que acredito serem importantes são:
- Impressões;
- Clicks;
- Características do anúncio (cores, disposição das palavras, organização, credibilidade)

Além disso, indo mais além, posso citar também a questão do site da empresa, um site que seja confiável, fácil de utilizar e que seja organizado, tem muitas mais chances de ter conversões.

No dataset repassado para mim, acredito que não pausaria nenhuma campanha, pois, correlacionando com a questão anterior, algumas campanhas geraram poucas conversões para a empresa, porém trouxeram boa receita, sem contar que atingiram seus objetivos, trazendo novos clientes ou gerando mais recompras de antigos clientes, e campanhas que não trouxeram uma boa receita, trouxeram um valor bom de conversões e impressões, desta forma, não teria como pausar nenhuma campanha, mas sim reformulá-las, modificando o anúncio, trocando cores, analisando um anúncio que teve mais clicks e observando a diferença entre eles para tentar equiparar e gerar uma paridade de impressões e clicks, para tentar gerar mais conversões e mais impressões, podendo trazer mais lucro para a empresa. E, por fim, ressalto a necessidade de analisar não somente uma métrica em conjunto, mas várias métricas juntas para ter uma análise do todo e ajudando a ter uma ideia melhor do que pode ser feito para melhorar a campanha e conversões da empresa.

### 3. Dentre as campanhas apresentadas, qual delas apresenta o maior potencial para escalar o investimento?

Analisando todo o dataset que foi disponibilizado, eu escolheria a campanha C para aumentar o investimento nela. Os motivos de eu escolher a C são a quantidade grande de impressões, sendo a campanha que mais teve impressões, totalizando 3.500.000 milhões, a grande quantidade de clicks na campanha também e um número que agrada é a métrica de conversões, que se destacou entre as campanhas e obteve o maior número. Também pode-se considerar a relação do gasto com a receita, pois a campanha dobrou a receita em relação ao gasto, assim conclui-se que a campanha teve um retorno positivo.

Ademais, essa campanha seria a mais viável para potencializar o investimento, porém, deve-se ficar atento também à outras campanhas, como a G e a K, que foram campanhas com investimentos baixos, porém que retornaram bons valores, tanto de receita para empresa como de visualizações (impressões e clicks). Desta forma, acredito que o investimento em potencial poderia ser dividido e remanejado para melhorar as campanhas em grupo, pois várias campanhas gerando bons resultados é melhor que uma campanha, mesmo que seja em potencial, gerando bons resultados. Pode-se gerar, desta maneira, uma segurança para a empresa em relação aos anúncios, pois se um anúncio parar de render números agradáveis, as outras campanhas podem conter e suprir a demanda desta que está em déficit, porém se existir uma só campanha com o investimento centralizado, caso esta campanha venha a tomar um rumo negativo, pode ser complicado para a empresa, pois terá que repensar e remanejar o investimento novamente. 

Concluo, portanto, que o investimento deve ser pensado cautelosamente, de forma a garantir a segurança da empresa relacionado à campanhas de anúncio. Porém, acredito que a campanha C seja uma grande oportunidade para aumentar a visibilidade e lucros da empresa, junta com as campanhas G e K.